PurchasesCoreSwift/Networking/HTTPClient.swift

```diff
@@ -0,0 +1,314 @@
+                }
+            }
+
+            if shouldBeginNextRequestWhenFinished {
+                var maybeNextRequest: HTTPRequest?
+                let logMessage = String(format: Strings.network.serial_request_done,
+                                        self.currentSerialRequest?.httpMethod ?? "",
+                                        self.currentSerialRequest?.path ?? "",
+                                        self.queuedRequests.count)
+                Logger.debug(logMessage)
+                self.currentSerialRequest = nil
+                if !self.queuedRequests.isEmpty {
+                    maybeNextRequest = self.queuedRequests[0]
+                    self.queuedRequests.remove(at: 0)
+                }
+                if let nextRequest = maybeNextRequest {
+                    Logger.debug(String(format: Strings.network.starting_next_request, nextRequest))
+                    self.performRequest(nextRequest.httpMethod,
+                                        performSerially: true,
+                                        path: nextRequest.path,
+                                        requestBody: nextRequest.requestBody,
+                                        headers: nextRequest.headers,
+                                        completionHandler: nextRequest.completionHandler)
+                }
+            }
+        }
+    }
+
+    func createRequest(httpMethod: String,
+                       path: String,
+                       requestBody maybeRequestBody: [AnyHashable: Any]?,
+                       headers: [String: String],
+                       refreshETag: Bool) -> URLRequest? {
+        let relativeURLString = "/v1\(path)"
+        guard let requestURL = URL(string: relativeURLString, relativeTo: SystemInfo.serverHostURL) else {
+            return nil
+        }
+
+        var request = URLRequest(url: requestURL)
+        request.httpMethod = httpMethod
+
+        let eTagHeader = eTagManager.eTagHeader(for: request, refreshETag: refreshETag)
+        let headersWithETag = headers.merging(eTagHeader) { (_, last) -> String in
+            last
+        }
+
+        request.allHTTPHeaderFields = headersWithETag
+
+        if httpMethod == "POST" {
+            if let requestBody = maybeRequestBody {
+                var jsonParseError: Error?
+                let isValidJSONObject = JSONSerialization.isValidJSONObject(requestBody)
+                if isValidJSONObject {
+                    do {
+                        request.httpBody =
+                            try JSONSerialization.data(withJSONObject: requestBody)
+                    } catch let error {
+                        jsonParseError = error
+                    }
+                }
+                if !isValidJSONObject || jsonParseError != nil {
+                    Logger.error(String(format: Strings.network.creating_json_error, requestBody))
+                    return nil
+                }
+            }
+        }
```

Comment:
perhaps this can be simplified? 
- unifiying `if`s
- not creating the `jsonError`, instead only using it when it's caught
- having a separate error flow for `jsonError` and `invalidJsonObject` with maybe a different error message

```swift
if httpMethod == "POST",
   let requestBody = maybeRequestBody {
    if JSONSerialization.isValidJSONObject(requestBody) {
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: requestBody)
        } catch let error {
            // perhaps a new message, since this message isn't using the error at all?
            Logger.error(String(format: Strings.network.creating_json_error, requestBody))
            return nil
        }
    } else {
        Logger.error(String(format: Strings.network.creating_json_error, requestBody))
        return nil
    }
}
return request
```